File: VB/GetSearchLocationAdditionalInfo/Form1.vb
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraMap

Namespace GetSearchLocationAdditionalInfo

    Partial Public Class Form1
        Inherits Form

        Const yourAzureKey As String = "Your Azure key here."

        Private map As MapControl

        Private searchProvider As AzureSearchDataProvider

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            PrepareMap()

            AddHandler Me.searchProvider.SearchCompleted, AddressOf SearchProvider_SearchCompleted
        End Sub

        Private Sub SearchProvider_SearchCompleted(ByVal sender As Object, ByVal e As AzureSearchCompletedEventArgs)
            Dim result As SearchRequestResult = e.RequestResult
            If result.ResultCode = RequestResultCode.Success Then
                Dim regions As List(Of LocationInformation) = result.SearchResults
                For Each region As LocationInformation In regions
                    AddPushpin(region.Location)
                    If idx = addresses.Count Then map.ZoomToFitLayerItems()
                Next

                DisplayResults(e.RequestResult)
                asyncResult = Me.BeginInvoke(CType(AddressOf SearchAsync, DoSearch))
            End If

            If result.ResultCode = RequestResultCode.BadRequest Then tbResults.Text += "The Azure Search service does not work for this location."
        End Sub

        Private Sub search_Click(ByVal sender As Object, ByVal e As EventArgs)
            idx = 0
            asyncResult = Me.BeginInvoke(CType(AddressOf SearchAsync, DoSearch))
        End Sub

        Friend Delegate Sub DoSearch()

        Private asyncResult As IAsyncResult

        Private idx As Integer = 0

        Private addresses As List(Of String) = New List(Of String) From {"505 N. Brand Blvd, Glendale CA 91203, USA", "1111 N Brand Blvd, Glendale, CA 91202, USA", "300 N Brand Blvd, Glendale, CA 91203, USA"}

        Private Sub SearchAsync()
            Me.EndInvoke(asyncResult)
            If idx < addresses.Count Then searchProvider.Search(addresses(System.Math.Min(System.Threading.Interlocked.Increment(idx), idx - 1)))
        End Sub

        Private Sub AddPushpin(ByVal geoPoint As GeoPoint)
            Dim pin As MapPushpin = New MapPushpin()
            pin.Location = geoPoint
            Dim layer As VectorItemsLayer = CType(map.Layers(2), VectorItemsLayer)
            CType(layer.Data, MapItemStorage).Items.Add(pin)
        End Sub

        Private Sub DisplayResults(ByVal requestResult As SearchRequestResult)
            Dim resultList As New StringBuilder("")
            If requestResult.ResultCode = RequestResultCode.Success Then
                Dim resCounter As Integer = 1
                For Each resultInfo As LocationInformation In requestResult.SearchResults
                    resultList.Append([String].Format(Microsoft.VisualBasic.Constants.vbLf & " Result {0}:  " & Microsoft.VisualBasic.Constants.vbLf, resCounter))
                    resultList.Append([String].Format(resultInfo.DisplayName & Microsoft.VisualBasic.Constants.vbLf))
                    resultList.Append([String].Format("Geographical coordinates:  {0}", resultInfo.Location))
                    resultList.Append([String].Format(Microsoft.VisualBasic.Constants.vbLf & "______________________________" & Microsoft.VisualBasic.Constants.vbLf))
                    resCounter += 1
                Next
            End If

            tbResults.Text += resultList.ToString()
        End Sub

        Private Sub PrepareMap()
            ' Create a map control.
            map = New MapControl()
            ' Specify the map position on the form.           
            map.Dock = DockStyle.Fill
            ' Add the map control to the window.
            Me.Controls.Add(map)
            ' Bring the map to the front.
            map.BringToFront()
            ' Create an image tiles layer and add it to the map.
            Dim tilesLayer As ImageLayer = New ImageLayer()
            map.Layers.Add(tilesLayer)
            ' Create an information layer and add it to the map.
            Dim infoLayer As InformationLayer = New InformationLayer()
            map.Layers.Add(infoLayer)
            Dim items As VectorItemsLayer = New VectorItemsLayer()
            items.Data = New MapItemStorage()
            map.Layers.Add(items)
            ' Create an Azure data provider and specify the Azure key.
            Dim azureProvider As AzureMapDataProvider = New AzureMapDataProvider()
            tilesLayer.DataProvider = azureProvider
            azureProvider.Tileset = AzureTileset.BaseLabelsRoad Or AzureTileset.Imagery
            azureProvider.AzureKey = yourAzureKey
            ' Create an Azure search data provider and specify the Azure key.
            searchProvider = New AzureSearchDataProvider With {.GenerateLayerItems = False, .AzureKey = yourAzureKey}
            infoLayer.DataProvider = searchProvider
            map.SearchPanelOptions.Visible = False
        End Sub
    End Class
End Namespace

File: CS/GetSearchLocationAdditionalInfo/Form1.cs

using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraMap;

namespace GetSearchLocationAdditionalInfo {
    public partial class Form1 : Form {

        const string yourAzureKey = "Your Azure key here.";
        MapControl map;
        AzureSearchDataProvider searchProvider;
        public Form1() {
            InitializeComponent();
        }
        private void Form1_Load(object sender, EventArgs e) {
            PrepareMap();
            searchProvider.SearchCompleted += SearchProvider_SearchCompleted;
        }
        private void SearchProvider_SearchCompleted(object sender, AzureSearchCompletedEventArgs e) {

            SearchRequestResult result = e.RequestResult;
            if (result.ResultCode == RequestResultCode.Success) {
                List<LocationInformation> regions = result.SearchResults;
                foreach (LocationInformation region in regions) {
                    AddPushpin(region.Location);
                    if (idx == addresses.Count)
                        map.ZoomToFitLayerItems();
                }
                DisplayResults(e.RequestResult);
                asyncResult = this.BeginInvoke((DoSearch)SearchAsync);
            }

            if (result.ResultCode == RequestResultCode.BadRequest)
                tbResults.Text += "The Azure Search service does not work for this location.";
        }

        private void search_Click(object sender, EventArgs e) {
            idx = 0;
            asyncResult = this.BeginInvoke((DoSearch)SearchAsync);
        }

        delegate void DoSearch();
        IAsyncResult asyncResult;
        int idx = 0;
        List<string> addresses = new List<string> {"505 N. Brand Blvd, Glendale CA 91203, USA",
            "1111 N Brand Blvd, Glendale, CA 91202, USA", "300 N Brand Blvd, Glendale, CA 91203, USA" };
        void SearchAsync() {
            this.EndInvoke(asyncResult);
            if (idx < addresses.Count)
                searchProvider.Search(addresses[idx++]);
        }

        private void AddPushpin(GeoPoint geoPoint) {
            MapPushpin pin = new MapPushpin();
            pin.Location = geoPoint;
            VectorItemsLayer layer = (VectorItemsLayer)this.map.Layers[2];
            ((MapItemStorage)layer.Data).Items.Add(pin);
        }

        private void DisplayResults(SearchRequestResult requestResult) {
            StringBuilder resultList = new StringBuilder("");

            if (requestResult.ResultCode == RequestResultCode.Success) {
                int resCounter = 1;
                foreach (LocationInformation resultInfo in requestResult.SearchResults) {
                    resultList.Append(String.Format("\n Result {0}:  \n", resCounter));
                    resultList.Append(String.Format(resultInfo.DisplayName + "\n"));
                    resultList.Append(String.Format("Geographical coordinates:  {0}", resultInfo.Location));
                    resultList.Append(String.Format("\n______________________________\n"));
                    resCounter++;
                }
            }
            tbResults.Text += resultList.ToString();
        }

        private void PrepareMap() {
            // Create a map control.
            map = new MapControl();

            // Specify the map position on the form.           
            map.Dock = DockStyle.Fill;

            // Add the map control to the window.
            this.Controls.Add(map);

            // Bring the map to the front.
            map.BringToFront();

            // Create an image tiles layer and add it to the map.
            ImageLayer tilesLayer = new ImageLayer();
            map.Layers.Add(tilesLayer);

            // Create an information layer and add it to the map.
            InformationLayer infoLayer = new InformationLayer();

            map.Layers.Add(infoLayer);

            VectorItemsLayer items = new VectorItemsLayer();
            items.Data = new MapItemStorage();
            map.Layers.Add(items);

            // Create an Azure data provider and specify the Azure key.
            AzureMapDataProvider azureProvider = new AzureMapDataProvider();
            tilesLayer.DataProvider = azureProvider;
            azureProvider.Tileset = AzureTileset.BaseLabelsRoad | AzureTileset.Imagery;
            azureProvider.AzureKey = yourAzureKey;

            // Create an Azure search data provider and specify the Azure key.
            searchProvider = new AzureSearchDataProvider { GenerateLayerItems = false, AzureKey = yourAzureKey };
            infoLayer.DataProvider = searchProvider;
            map.SearchPanelOptions.Visible = false;

        }
    }
}